File: modules/rust_workloads/src/fragmentation.rs
//! Fragmentation: allocate many variably sized blocks, free every other one,
//! then grow each survivor — live bytes shrink while the allocator's mapped
//! memory stays high.

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};

pub struct FragStats {
    pub survivors: usize,
    pub live_bytes: usize,
}

/// Returns the surviving blocks (hold them alive while the heap profile is
/// dumped) and their stats.
pub fn fragment(blocks: usize, seed: u64) -> (Vec<Vec<u8>>, FragStats) {
    let mut rng = StdRng::seed_from_u64(seed);
    let all: Vec<Vec<u8>> = (0..blocks)
        .map(|_| vec![0xA5_u8; rng.random_range(512..8192)])
        .collect();

    let mut survivors: Vec<Vec<u8>> = all
        .into_iter()
        .enumerate()
        .filter_map(|(i, block)| (i % 2 == 0).then_some(block))
        .collect();
    for block in &mut survivors {
        let target = block.len() * 2;
        block.resize(target, 0x5A);
    }

    let stats = FragStats {
        survivors: survivors.len(),
        live_bytes: survivors.iter().map(Vec::len).sum(),
    };
    (survivors, stats)
}

File: modules/cpp_workloads/src/fragmentation.cpp
#include "fragmentation.h"

#include <cstddef>
#include <cstdint>
#include <random>
#include <utility>
#include <vector>

namespace cpp_workloads {

auto fragment(std::size_t blocks, std::uint64_t seed)
    -> std::pair<std::vector<std::vector<std::byte>>, FragStats> {
    std::mt19937_64 rng(seed);
    std::uniform_int_distribution<std::size_t> size_dist(512, 8191);

    std::vector<std::vector<std::byte>> all;
    all.reserve(blocks);
    for (std::size_t i = 0; i < blocks; i++) {
        all.emplace_back(size_dist(rng), std::byte{0xA5});
    }

    std::vector<std::vector<std::byte>> survivors;
    survivors.reserve((blocks + 1) / 2);
    for (std::size_t i = 0; i < all.size(); i++) {
        if (i % 2 == 0) {
            survivors.push_back(std::move(all[i]));
        }
    }
    all.clear();
    all.shrink_to_fit();
    for (auto& block : survivors) {
        block.resize(block.size() * 2, std::byte{0x5A});
    }

    FragStats stats{.survivors = survivors.size(), .live_bytes = 0};
    for (const auto& block : survivors) {
        stats.live_bytes += block.size();
    }
    return {std::move(survivors), stats};
}

}  // namespace cpp_workloads
